// react-clock/build/Clock.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _mergeClassNames = require('merge-class-names');

var _mergeClassNames2 = _interopRequireDefault(_mergeClassNames);

var _Hand = require('./Hand');

var _Hand2 = _interopRequireDefault(_Hand);

var _Mark = require('./Mark');

var _Mark2 = _interopRequireDefault(_Mark);

var _dates = require('./shared/dates');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Clock = function (_Component) {
  _inherits(Clock, _Component);

  function Clock() {
    _classCallCheck(this, Clock);

    return _possibleConstructorReturn(this, (Clock.__proto__ || Object.getPrototypeOf(Clock)).apply(this, arguments));
  }

  _createClass(Clock, [{
    key: 'renderFace',
    value: function renderFace() {
      var _props = this.props,
          renderMinuteMarks = _props.renderMinuteMarks,
          renderHourMarks = _props.renderHourMarks,
          minuteMarksLength = _props.minuteMarksLength,
          minuteMarksWidth = _props.minuteMarksWidth,
          hourMarksLength = _props.hourMarksLength,
          hourMarksWidth = _props.hourMarksWidth;


      var marks = [];

      if (renderMinuteMarks) {
        for (var i = 1; i <= 60; i += 1) {
          var isHour = renderHourMarks && !(i % 5);
          var width = isHour ? hourMarksWidth : minuteMarksWidth;
          var length = isHour ? hourMarksLength : minuteMarksLength;

          marks.push(_react2.default.createElement(_Mark2.default, {
            angle: i * 6,
            key: isHour ? 'hour_' + Math.floor(i / 5) : 'minute_' + i,
            length: length,
            name: isHour ? 'hour' : 'minute',
            width: width
          }));
        }
      } else if (renderHourMarks) {
        for (var _i = 1; _i <= 12; _i += 1) {
          marks.push(_react2.default.createElement(_Mark2.default, {
            angle: _i * 30,
            key: 'hour_' + _i,
            length: hourMarksLength,
            name: 'hour',
            width: hourMarksWidth
          }));
        }
      }

      return _react2.default.createElement(
        'div',
        { className: 'react-clock__face' },
        marks
      );
    }
  }, {
    key: 'renderHourHand',
    value: function renderHourHand() {
      var _props2 = this.props,
          hourHandWidth = _props2.hourHandWidth,
          hourHandLength = _props2.hourHandLength,
          hourHandOppositeLength = _props2.hourHandOppositeLength,
          value = _props2.value;


      var angle = value ? (0, _dates.getHours)(value) * 30 + (0, _dates.getMinutes)(value) / 2 + (0, _dates.getSeconds)(value) / 600 : 0;

      return _react2.default.createElement(_Hand2.default, {
        angle: angle,
        name: 'hour',
        length: hourHandLength,
        oppositeLength: hourHandOppositeLength,
        width: hourHandWidth
      });
    }
  }, {
    key: 'renderMinuteHand',
    value: function renderMinuteHand() {
      if (!this.props.renderMinuteHand) {
        return null;
      }

      var _props3 = this.props,
          minuteHandWidth = _props3.minuteHandWidth,
          minuteHandLength = _props3.minuteHandLength,
          minuteHandOppositeLength = _props3.minuteHandOppositeLength,
          value = _props3.value;


      var angle = value ? (0, _dates.getHours)(value) * 360 + (0, _dates.getMinutes)(value) * 6 + (0, _dates.getSeconds)(value) / 10 : 0;

      return _react2.default.createElement(_Hand2.default, {
        angle: angle,
        name: 'minute',
        length: minuteHandLength,
        oppositeLength: minuteHandOppositeLength,
        width: minuteHandWidth
      });
    }
  }, {
    key: 'renderSecondHand',
    value: function renderSecondHand() {
      if (!this.props.renderSecondHand) {
        return null;
      }

      var _props4 = this.props,
          secondHandWidth = _props4.secondHandWidth,
          secondHandLength = _props4.secondHandLength,
          secondHandOppositeLength = _props4.secondHandOppositeLength,
          value = _props4.value;


      var angle = value ? (0, _dates.getMinutes)(value) * 360 + (0, _dates.getSeconds)(value) * 6 : 0;

      return _react2.default.createElement(_Hand2.default, {
        name: 'second',
        angle: angle,
        length: secondHandLength,
        oppositeLength: secondHandOppositeLength,
        width: secondHandWidth
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props5 = this.props,
          size = _props5.size,
          value = _props5.value;


      return _react2.default.createElement(
        'time',
        {
          className: (0, _mergeClassNames2.default)('react-clock', this.props.className),
          dateTime: value instanceof Date ? value.toISOString() : value,
          style: {
            width: size + 'px',
            height: size + 'px'
          }
        },
        this.renderFace(),
        this.renderHourHand(),
        this.renderMinuteHand(),
        this.renderSecondHand()
      );
    }
  }]);

  return Clock;
}(_react.Component);

exports.default = Clock;


Clock.defaultProps = {
  hourHandLength: 50,
  hourHandWidth: 4,
  hourMarksLength: 10,
  hourMarksWidth: 3,
  minuteHandLength: 70,
  minuteHandWidth: 2,
  minuteMarksLength: 6,
  minuteMarksWidth: 1,
  renderHourMarks: true,
  renderMinuteHand: true,
  renderMinuteMarks: true,
  renderSecondHand: true,
  secondHandLength: 90,
  secondHandWidth: 1,
  size: 150
};

Clock.propTypes = {
  className: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.arrayOf(_propTypes2.default.string)]),
  hourHandLength: _propTypes2.default.number,
  hourHandOppositeLength: _propTypes2.default.number,
  hourHandWidth: _propTypes2.default.number,
  hourMarksLength: _propTypes2.default.number,
  hourMarksWidth: _propTypes2.default.number,
  minuteHandLength: _propTypes2.default.number,
  minuteHandOppositeLength: _propTypes2.default.number,
  minuteHandWidth: _propTypes2.default.number,
  minuteMarksLength: _propTypes2.default.number,
  minuteMarksWidth: _propTypes2.default.number,
  renderHourMarks: _propTypes2.default.bool,
  renderMinuteHand: _propTypes2.default.bool,
  renderMinuteMarks: _propTypes2.default.bool,
  renderSecondHand: _propTypes2.default.bool,
  secondHandLength: _propTypes2.default.number,
  secondHandOppositeLength: _propTypes2.default.number,
  secondHandWidth: _propTypes2.default.number,
  size: _propTypes2.default.number,
  value: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.instanceOf(Date)])
};
